lychee-lib/src/checker/file.rs

```diff
@@ -110,35 +113,34 @@ impl FileChecker {
         }
     }
 
-    /// Checks if the given path exists and performs additional checks if necessary.
-    ///
-    /// First, the given path is resolved to a file by applying fallback extensions
-    /// and finding index files if needed. Then, the file is checked to make sure it
-    /// exists and passes any additional checks.
+    /// Resolves the given local path by applying logic which is specific to local file
+    /// checking - currently, this includes fallback extensions and index files.
     ///
     /// # Arguments
     ///
-    /// * `path` - The path to check.
-    /// * `uri` - The original URI, used for checking and error reporting.
+    /// * `path` - The path to check. Need not exist.
+    /// * `uri` - The original URI, used for error reporting.
     ///
     /// # Returns
     ///
-    /// Returns a `Status` indicating the result of the check.
-    async fn check_path(&self, path: &Path, uri: &Uri) -> Status {
+    /// Returns `Ok` with the resolved path if it is valid, otherwise returns
+    /// `Err` with an appropriate error. The returned path, if any, is guaranteed
+    /// to exist and may be a file or a directory.
+    fn resolve_local_path(&self, path: &Path, uri: &Uri) -> Result<PathBuf, ErrorKind> {
         let path = match path.metadata() {
             // for non-existing paths, attempt fallback extensions
             Err(e) if e.kind() == std::io::ErrorKind::NotFound => {
-                self.apply_fallback_extensions(path, uri).map(Cow::Owned)
+                self.apply_fallback_extensions(path, uri)
             }
 
             // other IO errors are unexpected and should fail the check
             Err(e) => Err(ErrorKind::ReadFileInput(e, path.to_path_buf())),
 
             // existing directories are resolved via index files
-            Ok(meta) if meta.is_dir() => self.apply_index_files(path).map(Cow::Owned),
+            Ok(meta) if meta.is_dir() => self.apply_index_files(path),
 
             // otherwise, path is an existing file - just return the path
```

Comment:
Have you tried to return a `Path` (with a lifetime)? This would technically be a breaking change, but that should be fine if we can avoid the copy.

Reply:
Silly me, you're right.

I played around with it and created https://github.com/rina-forks/lychee/pull/1. Can you take a look? I can see advantages and disadvantages on both sides, so I'm okay with keeping your version as well. 😉 If you decide to merge my PR, this PR should automatically get updated. 🪄

Reply:
It is merged, thanks for your help!